src/sbn_probability.cpp

```diff
@@ -232,25 +236,32 @@ EigenVectorXd SBNProbability::ExpectationMaximization(
       // Loop over the various rooting positions of this topology, using log_q_weights
       // to store the probability of the tree in the various rootings (we will normalize
       // it later).
+      double log_q_weight_sum = DOUBLE_NEG_INF;
+      NumericalUtils::ReportFloatingPointEnvironmentExceptions("|Before Rooting|");
       for (size_t rooting_position = 0; rooting_position < edge_count;
            ++rooting_position) {
         const SizeVector& rooted_representation =
             indexer_representation[rooting_position];
         // Calculate the SBN probability of this topology rooted at this position.
-        double log_p_rooted_topology = SumOf(sbn_parameters, rooted_representation, 0.);
-        log_q_weights[rooting_position] =
-            (log_p_rooted_topology > LOG_EPS) ? log_p_rooted_topology : LOG_EPS;
+        double log_p_rooted_topology = SumOfLogProbabilities(sbn_parameters,
+                                                     rooted_representation, 0.);
+        // SHJ: Sometimes overflow is reported, sometimes it's underflow...
+        if (fetestexcept(FE_OVERFLOW) || fetestexcept(FE_UNDERFLOW)) {
+          log_q_weights[rooting_position] = DOUBLE_MINIMUM;
+          feclearexcept(FE_OVERFLOW);
+          feclearexcept(FE_UNDERFLOW);
+        } else {
+          log_q_weights[rooting_position] = log_p_rooted_topology;
+        }
       }  // End of looping over rooting positions.
-      // This is the likelihood of the topology, marginalized over rooting position.
-      double log_p_unrooted_topology = NumericalUtils::LogSum(log_q_weights);
-      // Increase the total log marginal likelihood by this topologies contribution.
-      score_history[em_idx] += topology_count * log_p_unrooted_topology;
+      NumericalUtils::ReportFloatingPointEnvironmentExceptions("|After Rooting|");
+      log_q_weight_sum = NumericalUtils::LogSum(log_q_weights);
```

Comment:
I feel like the `log_p_unrooted_topology` that was in the previous version is actually a more descriptive name. What do you think?

Reply:
Done: log_q_weight_sum changed back to log_p_unrooted_topology.